mscore/timeline.cpp

```diff
@@ -0,0 +1,1522 @@
+
+                  std::pair<QGraphicsItem*, int> key1(key, 2);
+                  std::pair<QGraphicsItem*, int> key2(text, 2);
+                  meta_rows.push_back(key1);
+                  meta_rows.push_back(key2);
+                  }
+            }
+
+      for (s = _score->firstSegment(SegmentType::KeySig); s; s = s->next1(SegmentType::KeySig)) {
+            const KeySig* k = toKeySig(s->element(0));
+            int staff_track = 0;
+            while (!k && staff_track < rows) {
+                  staff_track++;
+                  k = toKeySig(s->element(staff_track * VOICES));
+                  }
+            if (!k) continue;
+            if (k->generated()) continue;
+
+            KeySig* nk = new KeySig(*k);
+            QList<Staff*> staves = _score->staves();
+
+            if (!nk->isAtonal() && !nk->isCustom() && !_score->styleB(StyleIdx::concertPitch)) {
+                  const Key key = k->key();
+                  const Interval firstInterval = staves.at(staff_track)->part()->instrument()->transpose();
+                  Key nKey = transposeKey(key, firstInterval);
+                  nk->setKey(nKey);
+                  }
+
+            int track = 0;
+            bool same = true;
+            for (Staff* st: staves) {
+                  if (!st->isPitchedStaff(s->tick())) {
+                        track += VOICES;
+                        continue;
+                        }
+                  const KeySig* curr = toKeySig(s->element(track));
+                  if (!curr) {
+                        const Interval curr_interval = st->part()->instrument()->transpose();
+                        Key tmp = Key::C;
+                        tmp = transposeKey(tmp, curr_interval);
+                        if (nk->key() == tmp) {
+                              track += VOICES;
+                              continue;
+                              }
+                        same = false;
+                        break;
+                        }
+                  KeySig* ncurr = new KeySig(*curr);
+                  //Transpose to concert pitch
+                  if (!ncurr->isAtonal() && !ncurr->isCustom() && !_score->styleB(StyleIdx::concertPitch)) {
+                        const Key key  = ncurr->key();
+                        const Interval firstInterval = st->part()->instrument()->transpose();
+                        Key nKey = transposeKey(key, firstInterval);
+                        ncurr->setKey(nKey);
+                        }
+                  if (*ncurr == *nk) {
+                        track += VOICES;
+                        continue;
+                        }
+                  same = false;
+                  break;
+                  }
+            if (!same) continue;
+
+            int keyInt = static_cast<int>(nk->key());
+
+            QString keys;
+            if (nk->isAtonal())
+                  keys = "X";
+            else if (nk->isCustom())
+                  keys = tr("Cu...");
```

Comment:
all the standard key sigs are going to two chars long max.  tbh, I can't think of the best way to deal with custom key sig names in this context, but I had suggested eliding as a quick fix for the time being since is rare...but you guys please make a better call.  I believe eventually most things in the timeline might have a tooltip anyway.